EKF/mag_fusion.cpp

```diff
@@ -47,51 +48,77 @@
-	// check for a badly conditioned covariance matrix
 	const char* numerical_error_covariance_reset_string = "numerical error - covariance reset";
-
+	const float HK0 = magE*q3;
+	const float HK1 = magN*q0;
+	const float HK2 = magD*q2;
+	const float HK3 = HK0 + HK1 - HK2;
+	const float HK4 = 2*HK3;
+	const float HK5 = magD*q3 + magE*q2 + magN*q1;
+	const float HK6 = 2*HK5;
+	const float HK7 = magE*q1;
+	const float HK8 = magD*q0;
+	const float HK9 = magN*q2;
+	const float HK10 = magD*q1;
+	const float HK11 = magE*q0;
+	const float HK12 = magN*q3;
+	const float HK13 = HK10 + HK11 - HK12;
+	const float HK14 = 2*HK13;
+	const float HK15 = ecl::powf(q1, 2);
+	const float HK16 = ecl::powf(q2, 2);
+	const float HK17 = -HK16;
+	const float HK18 = ecl::powf(q0, 2);
+	const float HK19 = ecl::powf(q3, 2);
+	const float HK20 = HK18 - HK19;
+	const float HK21 = HK15 + HK17 + HK20;
+	const float HK22 = q0*q3;
+	const float HK23 = q1*q2;
+	const float HK24 = HK22 + HK23;
+	const float HK25 = q1*q3;
+	const float HK26 = q0*q2;
+	const float HK27 = 2*HK24;
+	const float HK28 = -2*HK25 + 2*HK26;
+	const float HK29 = 2*HK5;
+	const float HK30 = 2*HK3;
+	const float HK31 = -HK7 + HK8 + HK9;
+	const float HK32 = 2*HK31;
+	const float HK33 = HK32*P(0,2);
+	const float HK34 = 2*HK13;
+	const float HK35 = HK34*P(0,3);
+	const float HK36 = HK21*P(0,16) + HK27*P(0,17) - HK28*P(0,18) + HK29*P(0,1) + HK30*P(0,0) - HK33 + HK35 + P(0,19);
+	const float HK37 = HK21*P(16,16) + HK27*P(16,17) - HK28*P(16,18) + HK29*P(1,16) + HK30*P(0,16) - HK32*P(2,16) + HK34*P(3,16) + P(16,19);
+	const float HK38 = HK21*P(16,18) + HK27*P(17,18) - HK28*P(18,18) + HK29*P(1,18) + HK30*P(0,18) - HK32*P(2,18) + HK34*P(3,18) + P(18,19);
+	const float HK39 = HK29*P(1,2);
+	const float HK40 = HK30*P(0,2);
+	const float HK41 = HK21*P(2,16) + HK27*P(2,17) - HK28*P(2,18) - HK32*P(2,2) + HK34*P(2,3) + HK39 + HK40 + P(2,19);
+	const float HK42 = HK21*P(16,17) + HK27*P(17,17) - HK28*P(17,18) + HK29*P(1,17) + HK30*P(0,17) - HK32*P(2,17) + HK34*P(3,17) + P(17,19);
+	const float HK43 = HK29*P(1,3);
+	const float HK44 = HK30*P(0,3);
+	const float HK45 = HK21*P(3,16) + HK27*P(3,17) - HK28*P(3,18) - HK32*P(2,3) + HK34*P(3,3) + HK43 + HK44 + P(3,19);
+	const float HK46 = HK32*P(1,2);
+	const float HK47 = HK34*P(1,3);
+	const float HK48 = HK21*P(1,16) + HK27*P(1,17) - HK28*P(1,18) + HK29*P(1,1) + HK30*P(0,1) - HK46 + HK47 + P(1,19);
+	const float HK49 = HK21*P(16,19) + HK27*P(17,19) - HK28*P(18,19) + HK29*P(1,19) + HK30*P(0,19) - HK32*P(2,19) + HK34*P(3,19) + P(19,19);
+
+	_mag_innov_var(0) = (HK21*HK37 + HK27*HK42 - HK28*HK38 + HK29*HK48 + HK30*HK36 - HK32*HK41 + HK34*HK45 + HK49 + R_MAG);
+	float HK50;
```

Comment:
```suggestion
	if (_mag_innov_var(0) < R_MAG) {
		// the innovation variance contribution from the state covariances is negative which means the covariance matrix is badly conditioned
		_fault_status.flags.bad_mag_x = true;
		// we need to re-initialise covariances and abort this fusion step
		resetMagRelatedCovariances();
		ECL_ERR("magX %s", numerical_error_covariance_reset_string);
		return;
	}
	_fault_status.flags.bad_mag_x = false;
	const float HK50 = 1.0F/_mag_innov_var(0);
```
The same applies to the other axes

Reply:
Done